File: go/examples/stk_push.go
// Example shows how to use the M-Pesa Go SDK to initiate an STK Push.
package main

import (
	"context"
	"fmt"
	"log"
	"os"

	"github.com/yourdudeken/mpesa-sdk/client"
	"github.com/yourdudeken/mpesa-sdk/types"
)

func main() {
	mpesa := client.NewClient(types.MpesaConfig{
		ConsumerKey:    os.Getenv("MPESA_CONSUMER_KEY"),
		ConsumerSecret: os.Getenv("MPESA_CONSUMER_SECRET"),
		Environment:    types.Sandbox,
		Passkey:        os.Getenv("MPESA_PASSKEY"),
	})

	ctx := context.Background()

	resp, err := mpesa.STKPush(ctx, types.STKPushRequest{
		BusinessShortCode: 174379,
		TransactionType:   types.CustomerPayBillOnline,
		Amount:            1,
		PartyA:            254722000000,
		PartyB:            174379,
		PhoneNumber:       254722111111,
		CallBackURL:       "https://example.com/callback",
		AccountReference:  "INV-001",
		TransactionDesc:   "Payment",
	})

	if err != nil {
		log.Fatalf("STK Push failed: %v", err)
	}

	fmt.Printf("Checkout ID: %s\n", resp.CheckoutRequestID)
}

File: typescript/src/services/stk-push.ts
import { MpesaApiClient } from "../client/client.js";
import { SANDBOX_ENDPOINTS } from "../environment.js";
import { generatePassword, generateTimestamp, Validation } from "../utils/index.js";
import { ValidationError } from "../errors/index.js";
import type {
  STKPushRequest,
  STKPushResponse,
  STKQueryRequest,
  STKQueryResponse,
  STKCallbackPayload,
  STKCallbackResult,
  TransactionType,
} from "../types/index.js";

const VALID_TRANSACTION_TYPES: readonly TransactionType[] = [
  "CustomerPayBillOnline",
  "CustomerBuyGoodsOnline",
] as const;

export class STKPushService {
  constructor(private readonly client: MpesaApiClient) {}

  async initiate(request: STKPushRequest): Promise<STKPushResponse> {
    const passkey = this.client.getConfig().passkey;

    if (!passkey) {
      throw new ValidationError("Passkey is required for STK Push.");
    }

    Validation.positiveNumber(request.BusinessShortCode, "BusinessShortCode");
    Validation.oneOf(request.TransactionType, "TransactionType", VALID_TRANSACTION_TYPES);
    Validation.amount(request.Amount, "Amount");
    Validation.phoneNumber(request.PartyA, "PartyA");
    Validation.phoneNumber(request.PartyB, "PartyB");
    Validation.phoneNumber(request.PhoneNumber, "PhoneNumber");
    Validation.validUrl(request.CallBackURL, "CallBackURL");
    Validation.maxLength(request.AccountReference, "AccountReference", 12);
    Validation.maxLength(request.TransactionDesc, "TransactionDesc", 13);

    const timestamp = request.Timestamp || generateTimestamp();
    const password = request.Password || generatePassword(
      request.BusinessShortCode,
      passkey,
      timestamp,
    );

    const payload: STKPushRequest = {
      ...request,
      Password: password,
      Timestamp: timestamp,
    };

    return this.client.post<STKPushResponse>(
      SANDBOX_ENDPOINTS.STK_PUSH,
      payload,
    );
  }

  async query(request: STKQueryRequest): Promise<STKQueryResponse> {
    const passkey = this.client.getConfig().passkey;
    if (!passkey) {
      throw new ValidationError("Passkey is required for STK Query.");
    }

    const timestamp = request.Timestamp || generateTimestamp();
    const password = request.Password || generatePassword(
      request.BusinessShortCode,
      passkey,
      timestamp,
    );

    const payload: STKQueryRequest = {
      ...request,
      Password: password,
      Timestamp: timestamp,
    };

    return this.client.post<STKQueryResponse>(
      SANDBOX_ENDPOINTS.STK_QUERY,
      payload,
    );
  }

  static parseCallback(payload: STKCallbackPayload): STKCallbackResult {
    const callback = payload.Body.stkCallback;
    const result: STKCallbackResult = {
      success: callback.ResultCode === 0,
      merchantRequestId: callback.MerchantRequestID,
      checkoutRequestId: callback.CheckoutRequestID,
      resultCode: callback.ResultCode,
      resultDescription: callback.ResultDesc,
    };

    if (callback.CallbackMetadata?.Item) {
      for (const item of callback.CallbackMetadata.Item) {
        switch (item.Name) {
          case "Amount":
            result.amount = Number(item.Value);
            break;
          case "MpesaReceiptNumber":
            result.receiptNumber = String(item.Value);
            break;
          case "TransactionDate":
            result.transactionDate = String(item.Value);
            break;
          case "PhoneNumber":
            result.phoneNumber = String(item.Value);
            break;
        }
      }
    }

    return result;
  }
}
